spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/Mqttv5PahoMessageDrivenChannelAdapter.java

```diff
@@ -162,6 +204,11 @@ protected void onInit() {
 
 	@Override
 	protected void doStart() {
+		if (getClientManager() != null) {
+			subscribeToAll();
```

Comment:
Oh! I see now: you have broken the logic.
We must not try to subscribe in the `doStart()`.
We have to try to do that exactly from the `connectComplete()` callback.
The problem is present just because `mqttClient.reconnect()` is an async scheduled operation.
So, even if client manager is started, it doesn't mean that it has been connected.
Therefore we really need to wait for the `connectComplete()` before trying to subscribe.
Please, see my other comment where I suggest to implement `addConnectCompleteCallback()`

Reply:
> Since we cannot provide this `MqttCallback` to a shared client, we have to implement something like `addConnectCompleteCallback(BiConsumer<boolean, String> connectCompleteCallback`) in the ClientManager.


I already did smth similar but I have created an inner `Callback` interface in `ClientManager` with semantically clear method `connectComplete(boolean isReconnect, String serverURI)` - I guess it should be easier for users to understand the purpose of the callback when they see parameter and method names then just plain `BiConsumer` with types. If you don't have objections I would like to use this interface.

Reply:
Sure, we can go that way, but let’s name it a `ConnectCallback` to be more specific . Also I don’t see a reason in the `url` param: we definitely know to what manager we provide a callback.